rfcs/2021-09-14-3910-schema-support.md

```diff
@@ -0,0 +1,472 @@
+While we can optimize this as much as possible, the inherent cost of walking an
+event and doing comparisons cannot be ignored, and has to be accepted as an
+inherent cost of implementing schema support.
+
+As developers, we instinctively search for solutions to this problem, such as
+sampled schema validation, to limit the impact on performance, but because all
+internal systems will be built _assuming_ schema adherence is guaranteed, there
+can be no room for error, nor for shortcuts that give you mostly statistically
+correct results, but not always.
+
+Even so, aside from the impact on the internal systems, if we promote strict
+schema adherence, the expectation is that no badly formatted events will end up
+at the configured destination.
+
+Because of this fact, this RFC proposes one solution to start with: you either
+_enable_ schema support, and accept the inherent performance cost, or you
+_disable_ support, cannot be guaranteed any schema adherence, but keep the most
+amount of performance. In the future, alternative "strategies" might be added,
+to find a middle-ground, but this would likely require a new RFC, given the
+mentioned implications.
+
+### Invalid Events
+
+For any source that has a schema configured, it can either **drop** or
+**accept** invalid events.
+
+#### Dropping Events
+
+When a source is configured to drop invalid events, and an event does not match
+the schema, the event won't be allowed to proceed to the next component in the
+pipeline.
+
+When this happens, the event is dropped, and an internal event is triggered
+indicating the failed schema validation.
+
+#### Accept Invalid Events
+
+If the source is configured to accept invalid events, it internally exposes
+itself as exporting events matching both the operator-configured, and the "any"
+schema. This is the least-restrictive schema definition.
+
+Allowing invalid events to be ingested has consequences. For one, downstream
+components can no longer rely on strict schema enforcement (in a technical sense
+they can, since all events adhere to the "any" schema). For example, for
+implicit schema transformations, the "any" schema provides no event structure
+guarantees, and thus no transformation can be done from events matching this
+schema.
+
+On the surface, the "any" schema thus becomes a useless piece of information.
+However, operators can choose to re-route mismatched events, by adding a `route`
+transform in-between. This transform can route any events matching the more
+strict schema to one lane, and all other events matching the "any" schema to
+another lane.
+
+#### Dead-Letter Pipeline
+
+In the future, dropped events can be picked up by a [dead-letter pipeline][dlq],
+but this is out-of-scope for this RFC.
+
+In a sense, accepting invalid events, combined with the `route` transform,
+already lets you manually implement a dead-letter pipeline.
+
+[dlq]: https://github.com/vectordotdev/vector/issues/1772
```

Comment:
I think the work Luke is doing to enable transform dead letter queues should be landing soon so I think this RFC could just use that concept as the preferred approach (extending it to sources). I realize it is likely out-of-scope of this particular RFC, but I don't think we should implement source schema support without it only to quickly add it later and need to make a backwards incompatible change.

We could list it as a dependency for this RFC and have a separate RFC for it.